File: python/print.py
import json
from process import process

MAX_COLUMNS = 10

PAGE_TEMPLATE = '''<!doctype html>
<html>
    <head>
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Training Plan</title>
        <style>
            table {{
                border-collapse: collapse;
            }}
            th, td {{
                border: 1px solid black;
                font-size: 14px;
                text-align: left;
            }}
            .week {{
                height: 1.6em;
                font-weight: bold;
            }}
            .day {{
                height: 1.4em;
                font-weight: bolder;
            }}
            .set {{
                width: 2cm;
            }}
        </style>
    </head>
    <body>
        <div>
            <table>
                {0}
            </table>
        <div>
    </body>
</html>'''

def render(content):
    data = json.loads(content)
    rows = [
        '<tr><td class="week" colspan="11">{0}</td></tr>'.format(data['week']),
        '<tr>{0}<tr>'.format(format_weights(data['weights']))
    ];
    colspan = MAX_COLUMNS + 1
    for day in data['days']:
        rows.append('<tr><td colspan="{0}">&nbsp;</td></tr>'.format(colspan))
        rows.append('<tr><td class="day" colspan="{0}">{1}</td></tr>'.format(colspan, day['name']))
        for exercise in day['exercises']:
            sets = exercise['sets']
            format_cell = format_ratio_cell if 'ratio' in sets[0] else format_simple_cell
            rows.append(format_row(exercise['name'], sets, format_cell))
            if len(sets) > MAX_COLUMNS:
                rows.append(format_row('', sets[MAX_COLUMNS:], format_cell))
    return PAGE_TEMPLATE.format('\n'.join(rows))

def format_weights(weights):
    items = ['<td></td>']
    for item in weights:
        items.append('<td class="set">{name}</td><td class="set">{weight}</td>'.format(**item))
    return ''.join(items)

def format_ratio_cell(cell):
    return '<td class="set">{0}% {1} / {2}</td>'\
        .format(round(cell['ratio'] * 100), cell['count'], cell['weight'])

def format_simple_cell(cell):
    return '<td class="set">{0}</td>'.format(cell['count'])

def format_row(name, sets, format_cell):
    cells = ['<td class="exercise">{0}</td>'.format(name)]
    cells.extend(map(format_cell, sets[:MAX_COLUMNS]))
    for _ in range(len(sets), MAX_COLUMNS):
        cells.append('<td></td>')
    return '<tr>{0}</tr>'.format(''.join(cells))

process(render)

File: node/print.js
const readStdin = require('./stdin');

const MAX_COLUMNS = 10;

function formatPage(table) {
    return `<!doctype html>
<html>
    <head>
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Training Plan</title>
        <style>
            table {
                border-collapse: collapse;
            }
            th, td {
                border: 1px solid black;
                font-size: 14px;
                text-align: left;
            }
            .week {
                height: 1.6em;
                font-weight: bold;
            }
            .day {
                height: 1.4em;
                font-weight: bolder;
            }
            .set {
                width: 2cm;
            }
        </style>
    </head>
    <body>
        <div>
            <table>
${table}
            </table>
        <div>
    </body>
</html>`;
}

function formatRatioCell(cell) {
    return `<td class="set">${(cell.ratio * 100).toFixed(0)}% ${cell.count} / ${cell.weight}</td>`;
}

function formatSimpleCell(cell) {
    return `<td class="set">${cell.count}</td>`;
}

function formatRow(name, sets, formatCell) {
    const cells = sets.slice(0, MAX_COLUMNS).map(formatCell);
    for (let i = sets.length; i < MAX_COLUMNS; ++i) {
        cells.push('<td></td>');
    }
    const content = [`<td class="exercise">${name}</td>`].concat(cells).join('');
    return `<tr>${content}</tr>`;
}

function formatWeights(weights) {
    return [
        '<td></td>',
        ...weights.map(({ name, weight }) =>
            `<td class="set">${name}</td><td class="set">${weight}</td>`)
    ].join('');
    return columns.join('');
}

function print(content) {
    const { week, weights, days } = JSON.parse(content);
    const rows = [
        `<tr><td class="week" colspan="11">${week}</td></tr>`,
        `<tr>${formatWeights(weights)}<tr>`
    ];
    days.forEach((day) => {
        rows.push(`<tr><td colspan="${MAX_COLUMNS + 1}">&nbsp;</td></tr>`);
        rows.push(`<tr><td class="day" colspan="${MAX_COLUMNS + 1}">${day.name}</td></tr>`);
        day.exercises.forEach(({ name, sets }) => {
            const formatCell = sets[0].ratio ? formatRatioCell : formatSimpleCell;
            rows.push(formatRow(name, sets, formatCell));
            if (sets.length > MAX_COLUMNS) {
                rows.push(formatRow('', sets.slice(MAX_COLUMNS), formatCell));
            }
        });
    });
    return formatPage(rows.join('\n'));
}

readStdin().then((data) => {
    process.stdout.write(print(data));
});
